src/nvim/api/win_config.c

```diff
@@ -236,18 +341,203 @@ void nvim_win_set_config(Window window, Dict(float_config) *config, Error *err)
+        api_set_error(err, kErrorTypeValidation, "Cannot move last window");
+        return;
+      } else if (parent != NULL && parent->handle == win->handle) {
+        int n_frames = 0;
+        for (frame_T *fr = win->w_frame->fr_parent->fr_child; fr != NULL; fr = fr->fr_next) {
+          n_frames++;
+        }
+
+        win_T *neighbor = NULL;
+
+        if (n_frames > 2) {
+          // There are three or more windows in the frame, we need to split a neighboring window.
+          frame_T *frame = win->w_frame->fr_parent;
+
+          if (frame->fr_parent) {
+            //   ┌──────────────┐
+            //   │      A       │
+            //   ├────┬────┬────┤
+            //   │ B  │ C  │ D  │
+            //   └────┴────┴────┘
+            //          ||
+            //          \/
+            // ┌───────────────────┐
+            // │         A         │
+            // ├─────────┬─────────┤
+            // │         │    C    │
+            // │    B    ├─────────┤
+            // │         │    D    │
+            // └─────────┴─────────┘
+            if (fconfig.split == kWinSplitAbove || fconfig.split == kWinSplitLeft) {
+              neighbor = win->w_next;
+              while (neighbor->w_floating) {
+                neighbor = neighbor->w_next;
+              }
+            } else {
+              neighbor = win->w_prev;
+              while (neighbor->w_floating) {
+                neighbor = neighbor->w_prev;
+              }
+            }
+          }
+          // If the frame doesn't have a parent, the old frame
+          // was the root frame and we need to create a top-level split.
+          int dir;
+          new_curwin = winframe_remove(win, &dir, win_tp == curtab ? NULL : win_tp);
+        } else if (n_frames == 2) {
+          // There are two windows in the frame, we can just rotate it.
+          int dir;
+          neighbor = winframe_remove(win, &dir, win_tp == curtab ? NULL : win_tp);
+          new_curwin = neighbor;
+        } else {
+          // There is only one window in the frame, we can't split it.
+          api_set_error(err, kErrorTypeValidation, "Cannot split window into itself");
+          return;
+        }
+        // Set the parent to whatever the correct
+        // neighbor window was determined to be.
+        parent = neighbor;
+      } else {
+        int dir;
+        new_curwin = winframe_remove(win, &dir, win_tp == curtab ? NULL : win_tp);
+      }
+      // move to neighboring window if we're moving the current window to a new tabpage
+      if (curwin == win && parent != NULL && new_curwin != NULL
+          && win_tp != win_find_tabpage(parent)) {
+        win_enter(new_curwin, true);
+      }
+      // Remove the old window from the tree of frames,
+      win_remove(win, win_tp == curtab ? NULL : win_tp);
+    } else {
+      // Remove the old window from the tree of frames
+      win_remove(win, win_tp == curtab ? NULL : win_tp);
```

Comment:
What does `win_remove()` have to do with "tree of frames"?

Reply:
Ah I think that was just a copy paste mistake and was originally referencing `winframe_remove`.

I'll remove these. There's definitely a good bit of cleanup I have to do here, sorry about that.